### DISCUSSION TITLE
Enable pinDigest to convert a major version to semver

### DISCUSSION THREAD

**Discussion by suzuki-shunsuke** [renovatebot/renovate#21901] [Suggest an Idea]
### Type of discussion.

I'm proposing an idea

### Tell us more.

This is a feature request.

- https://docs.renovatebot.com/modules/manager/github-actions/#additional-information
- https://docs.renovatebot.com/presets-helpers/#helperspingithubactiondigests

Currently, pinDigest updates GitHub Actions version as the following.

AS IS

```yaml
- uses: actions/checkout@v3
```

TO BE

```yaml
- uses: actions/checkout@8e5e7e5ab8b370d6c329ec480221332ada57f0ab # v3
```

e.g. https://github.com/suzuki-shunsuke/test-renovate-2/pull/94

I'd like Renovate to convert the version to semver.

```yaml
- uses: actions/checkout@8e5e7e5ab8b370d6c329ec480221332ada57f0ab # v3.5.2
```

## Why?

Even if the tag is not semver Renovate will update the commit hash, but you can't understand what is changed from the Renovate's pull request.

e.g. https://github.com/suzuki-shunsuke/test-github-action/pull/141

<img width="937" alt="image" src="https://user-images.githubusercontent.com/13323303/231947080-75930df5-a471-4b1a-a5ab-de5a270d7738.png">

On the other hand, if the tag is semver, you can understand what is changed from the Renovate's pull request.

e.g. https://github.com/suzuki-shunsuke/test-github-action/pull/143

<img width="947" alt="image" src="https://user-images.githubusercontent.com/13323303/231948517-01fcbf19-9f6d-467a-9bb5-7cba097f2233.png">

You can see the release note and you can understand this is a patch update.

**Comment by suzuki-shunsuke** [comment]
I created a CLI to resolve this issue.

https://github.com/suzuki-shunsuke/pinact

But if Renovate supports this feature, we don't have to use this tool.
And this tool supports only GitHub Actions for now.

**Reply by MPV** [reply]
Nice to see! 

Just to share other options, here's another one I found (which even is a GitHub Action):
- https://github.com/mheap/pin-github-action

**Comment by rarkins** [comment]
I don't think this is related to pinDigest, instead it is related to _version_ pinning.

Pre-requisite: #21902 (Use pinVersions instead of rangeStrategy=pin)

Related feature for Docker: #1027 (Dockerfile FROM pinning support)

One challenge is that the SHA for e.g. `v3` may not be the same as the sha for e.g. `v3.5.1`.

@suzuki-shunsuke I have a question: do you think Renovate should "blindly" pin to v3.5.1 in the above case, or only do so if it can ensure that the SHA is the same?

**Reply by suzuki-shunsuke** [reply]
If there is a semver such as `v3.5.1` whose commit hash is same with `v3`, it is good to convert the tag to the semver.
On the other hand, if there is no semver whose commit hash is same with `v3`, there are two options.

1. Convert the tag to the most latest tag in tags `v3.x.x`
1. Keep the tag `v3`

It is difficult to decide which option is better, but I think the option 2 is better.
I expect this is a rare case.

**Reply by rarkins** [reply]
This might work ok for GitHub Actions where we get both the tag and sha for all tags and we can easily filter for tags matching the SHA. For Docker, it costs us O(n) for n tags so we want to avoid doing that.

**Reply by viceice** [reply]
you can already convert short action versions to long format

https://github.com/renovatebot/.github/blob/338858852accdade30ea259ce0a7a3d7efe6584b/default.json#L39

**Reply by suzuki-shunsuke** [reply]
> you can already convert short action versions to long format
> 
> https://github.com/renovatebot/.github/blob/338858852accdade30ea259ce0a7a3d7efe6584b/default.json#L39

Thank you for your answer! It works!

https://github.com/suzuki-shunsuke/test-renovate-2/blob/f44f0ebb0db4b478ce4bedb11ade3d41a66bdeae/renovate.json5#L9-L15

- https://github.com/suzuki-shunsuke/test-renovate-2/pull/95

Note that the release note is wrong because the commit hash of v3 is same with v3.5.2.

**Reply by MaxymVlasov** [reply]
>https://github.com/suzuki-shunsuke/test-renovate-2/pull/95Note that the release note is wrong because the commit hash of v3 is same with v3.5.2.
>Note that the release note is wrong because the commit hash of v3 is same with v3.5.2.

No one knows which version was used successfully last time. It can be anything from 3.0.0 to 3.5.2.
I prefer to see the worst-case scenario update notes, in case someone did violate semver and did breaking changes in MINOR or PATCH

**Reply by MaxymVlasov** [reply]
> You can already convert short action versions to long format
> [renovatebot/.github@3388588/default.json#L39](https://github.com/renovatebot/.github/blob/338858852accdade30ea259ce0a7a3d7efe6584b/default.json#L39)

@viceice @rarkins would you mind specifying this temporary solution (which already exist for 2,5 years) somewhere in docs?
Because I copy-pasted that solution on my day 1 of Renovate usage (7 months ago), but have not sure until today what exactly it's doing.

Or write a paragraph about "feel free to check out these repos/orgs to find config inspiration which deals with unusual things"?

**Reply by rarkins** [reply]
Docs only PRs are welcome

**Comment by MaxymVlasov** [comment]
@rarkins I am glad to add 

```json
{
  "packageRules": [
    {
      "description": "v prefix workaround for action updates",
      "matchDepTypes": ["action"],
      "extractVersion": "^(?<version>v\\d+\\.\\d+\\.\\d+)$",
      "versioning": "regex:^v(?<major>\\d+)(\\.(?<minor>\\d+)\\.(?<patch>\\d+))?$"
    }
  ]
}
```

to docs for https://docs.renovatebot.com/presets-helpers/#helperspingithubactiondigests here:

![image](https://github.com/renovatebot/renovate/assets/11096782/01df864c-1c9b-43d1-9d88-edef867bc708)

but find that there are no docs, and the full section generated from https://github.com/renovatebot/renovate/blob/c69247c9db9b6031ab17ea824bce53ce398094f7/lib/config/presets/internal/helpers.ts#L24-L31

Should I send it as another helper preset, or as a [workaround preset](https://docs.renovatebot.com/presets-workarounds/) + mention the workaround in `pinGitHubActionDigests` helper description?

**Reply by rarkins** [reply]
Docs for these are autogenerated from all `helpers:*` presets in the repo. I think the `description` field needs to be more descriptive though so that people can understand what it's for

**Reply by MaxymVlasov** [reply]
Here is a possible solution - https://github.com/renovatebot/renovate/pull/23577

**Reply by HonkingGoose** [reply]
I'm trying to get a new preset into the code, with this PR:

- https://github.com/renovatebot/renovate/pull/23663

**Comment by suzuki-shunsuke** [comment]
I close this discussion because the issue was solved by the preset `helpers:pinGitHubActionDigestsToSemver`.

- https://github.com/renovatebot/renovate/pull/23663

I've confirmed the preset works well.

- https://github.com/suzuki-shunsuke/test-renovate-2/pull/105

<img width="1522" alt="image" src="https://github.com/renovatebot/renovate/assets/13323303/0c4e910e-dfa1-4f53-98b7-4c3a089393f5">

Thank you, everyone!

**Reply by MPV** [reply]
I can confirm that `helpers:pinGitHubActionDigestsToSemver` works really well for upgrading versions to semver. 🎉 

However, I may have run into a follow-up issue. Doee anyone have ideas why the "pin digests" functionality doesn't combine with `helpers:pinGitHubActionDigestsToSemver`? 

See minimal reproduction repo and example here:
- https://github.com/MPV/renovate-pinGitHubActionDigestsToSemver-should-pinDigests-to-semver/pull/2

**Reply by HonkingGoose** [reply]
> I can confirm that `helpers:pinGitHubActionDigestsToSemver` works really well for upgrading versions to semver. 🎉

Hooray! :partying_face: 

> However, I may have run into a follow-up issue. Doee anyone have ideas why the "pin digests" functionality doesn't combine with `helpers:pinGitHubActionDigestsToSemver`?

I wish I could help, but I only bashed that preset together with _a lot of help_ from others. :upside_down_face: So I hope somebody else can review the reproduction, and help you. :wink:

**Reply by MPV** [reply]
Thanks. ❤️  

I raised a separate discussion, in case it's more useful or simplifies collaboration on the follow-up issue:
- https://github.com/renovatebot/renovate/discussions/27194

**Reply by MPV** [reply]
FYI: I raised a follow-up issue (as this preset broke down entirely after other changes in version detection in Renovate, and any GitHub Actions PRs were just autoclosed). 

I made a PR that hopefully should fix it, more details there: 
- https://github.com/renovatebot/renovate/pull/27453
